framework/src/modules/chain/dpos/delegates_list.js

```diff
@@ -53,33 +53,54 @@ class DelegatesList extends EventEmitter {
 		return shuffleDelegateListForRound(round, list);
 	}
 
-	async getDelegatePublicKeysSortedByVote() {
+	async getDelegatePublicKeysSortedByVoteWeight(tx) {
 		const filters = { isDelegate: true };
 		const options = {
 			limit: this.activeDelegates,
 			sort: ['voteWeight:desc', 'publicKey:asc'],
 		};
-		const accounts = await this.storage.entities.Account.get(filters, options);
+		const accounts = await this.storage.entities.Account.get(
+			filters,
+			options,
+			tx,
+		);
 		return accounts.map(account => account.publicKey);
 	}
 
 	async generateActiveDelegateList(round) {
-		let delegatePublicKeys = await this.storage.entities.RoundDelegates.getRoundDelegates(
+		const delegatePublicKeys = await this.storage.entities.RoundDelegates.getRoundDelegates(
 			round,
 		);
 
 		if (!delegatePublicKeys.length) {
-			delegatePublicKeys = await this.getDelegatePublicKeysSortedByVote();
-
-			await this.storage.entities.RoundDelegates.create({
-				round,
-				delegatePublicKeys,
-			});
+			throw new Error(`No delegate list found for round: ${round}`);
 		}
 
 		return delegatePublicKeys;
 	}
 
+	async createRoundDelegateList(round, tx) {
```

Comment:
It is still possible to call this function with a round number from future. I think it's better the fetch the latest record and check provided round number and eventually update or insert accordingly.

Also I suggest to rename this `generateActiveDelegateList` since it has kind of took over the current `generateActiveDelegateList`'s job. And as I mentioned above, move the logic in existing `generateActiveDelegateList` method into `getRoundDelegates`.

Reply:
Good catch, I don't think we are checking for this now. Also, we should add an extra test for that.

Reply:
In what case it would be called with a round from the future?

Reply:
@lsilvs Not in the current business logic. However, method should throw exception if a round from future was provided, just as a precaution.

Reply:
This method is not exposed outside Dpos module.
Adding a comment saying `createRoundDelegateList` should only be called after each round finishes is enough imo.